### PR TITLE
post-install.sh bug starts gitea as root not git

### PR DESCRIPTION
**Pull request by nvatvani**
gitea docker image should never be started as root user.
It should be started under git user as described in https://docs.gitea.io/en-us/install-with-docker/#basics

### CODE REVIEW

**Comment by EndMove** [comment]
Hi nvatvani, it is true that Gitea should not be run as root. Nevertheless I had no choice but to do it that way. Gitea needs access to the server's SSL certificate and since ADM 4.0 ssl.crt and ssl.pem are read only for guests but this is not the case for ssl.key which requires root access.

I will contact the Asustor developers soon to find out more about the permission problem.

**Comment by nvatvani** [comment]
> Hi nvatvani, it is true that Gitea should not be run as root. Nevertheless I had no choice but to do it that way. Gitea needs access to the server's SSL certificate and since ADM 4.0 ssl.crt and ssl.pem are read only for guests but this is not the case for ssl.key which requires root access.
> 
> I will contact the Asustor developers soon to find out more about the permission problem.

Hmm - that does not seem to be the case for my device:
admin@AS6404T-XXXX:/volume0/usr/builtin/etc/certificate $ ls -lart
total 40
drwxr-xr-x    3 root     root          4096 Apr  4  2020 letsencrypt/
-rw-r--r--    1 root     root          1675 Jul 25 00:34 ssl.key
-rw-r--r--    1 root     root          5638 Nov 23 00:00 ssl.crt
-rw-r--r--    1 root     root          7314 Nov 23 00:00 ssl.pem
drwxr-xr-x    5 root     root          4096 Dec  7 16:25 ssl/
-rw-r--r--    1 root     root           765 Dec 15 00:00 certificate.json

You could also fix the key yourself by doing "sudo chmod 644 /volume0/usr/builtin/etc/certificate/ssl.key" as admin user.

Sounds like checking/setting perms for the certs should be a part of the pre-install.

**Comment by EndMove** [comment]
> > Hi nvatvani, it is true that Gitea should not be run as root. Nevertheless I had no choice but to do it that way. Gitea needs access to the server's SSL certificate and since ADM 4.0 ssl.crt and ssl.pem are read only for guests but this is not the case for ssl.key which requires root access.
> > I will contact the Asustor developers soon to find out more about the permission problem.
> 
> Hmm - that does not seem to be the case for my device: admin@AS6404T-XXXX:/volume0/usr/builtin/etc/certificate $ ls -lart total 40 drwxr-xr-x 3 root root 4096 Apr 4 2020 letsencrypt/ -rw-r--r-- 1 root root 1675 Jul 25 00:34 ssl.key -rw-r--r-- 1 root root 5638 Nov 23 00:00 ssl.crt -rw-r--r-- 1 root root 7314 Nov 23 00:00 ssl.pem drwxr-xr-x 5 root root 4096 Dec 7 16:25 ssl/ -rw-r--r-- 1 root root 765 Dec 15 00:00 certificate.json
> 
> You could also fix the key yourself by doing "sudo chmod 644 /volume0/usr/builtin/etc/certificate/ssl.key" as admin user.
> 
> Sounds like checking/setting perms for the certs should be a part of the pre-install for gitea-adm

Hello nvatvani,
I'm glad to see that you are proposing solutions. I just ran several simulations and identified exactly where the permissions problem was. In all the classic use cases of the certificate manager, the permissions are correct (644).

#### Certificats Suport (asustor.com)
```
root@Dev :/volume0/usr/builtin/etc/certificate # ls -lr
total 24
-rwxr-xr-x    1 root     root          3111 Aug  5 07:51 ssl.pem*
-rwxr-xr-x    1 root     root          1675 Aug  5 07:51 ssl.key*
-rwxr-xr-x    1 root     root          1436 Aug  5 07:51 ssl.crt*
```

#### Imported certificat
```
root@Dev :/volume0/usr/builtin/etc/certificate # ls -lr
total 32
-rwxr-xr-x    1 root     root          7367 Jan  6 23:02 ssl.pem
-rwxr-xr-x    1 root     root          1692 Jan  6 23:02 ssl.key
-rwxr-xr-x    1 root     root          5674 Jan  6 23:02 ssl.crt
```

#### Let's Encrypt certificat
```
root@Dev :/volume0/usr/builtin/etc/certificate # ls -lr
total 32
-rw-r--r--    1 root     root          7402 Jan  6 23:24 ssl.pem
-rw-------    1 root     root          1704 Jan  6 23:24 ssl.key
-rw-r--r--    1 root     root          5697 Jan  6 23:24 ssl.crt
```

We can see here that the problem only occurs when using a Let's Encrypt certificate. This suggests that an mistake was made when the upgrade of Let's Encrypt for ADM 4.0.
Your solution of changing the permissions cannot be applied unfortunately. This is because it conflicts with ASUSTOR's guidelines for application creation. "An ADM application should not and cannot change any behaviour, configuration or system settings."

I can't neglect the Let's Encrypt users either... Gitea being reliable for the moment, the root user problem can wait until the permission problem is fixed.

Kind regards,
Jérémi.

**Comment by EndMove** [comment]
UP: I have just contacted Asustor and am waiting for a reply.

**Comment by nvatvani** [comment]
Nice - so while they give a response, there is a sensible solution to this problem.

If ssl.key is indeed not globally readable, post-install.sh can check for this. The same check should be extended to ssl.crt for consistency.

Only when both are indeed globally readable should 'git' user be used in docker config and HTTPS become the default.

If any of ssl.key or ssl.crt not be globally readable, 'git' user should still be used but HTTP is what gets used instead of HTTPS.

This fix in gitea-adm will ensure the proper usage of the docker image regardless of when/how/if asustor fixes the permissions.

**Comment by EndMove** [comment]
Hi nvatvani, la solution que tu proposes est intéressante. Je ne vais pas pouvoir travailler dessus avant plusieurs jours, néanmoins si elle te tien à cœur je me ferais un plaisir de merge ta solution ainsi que de te mentionner dans les contributeurs.

**Comment by EndMove** [comment]
Hi nvatvani, the solution you propose is interesting, but I won't be able to work on it for a few days. If this solution is important or urgent for you, I will be happy to merge your solution and mention you in the contributors.

**Comment by nvatvani** [comment]
No rush - you can take your time any maybe even wait for a response from Asustor Support.

I can only say that when I was on ADM 3, I used Gitea built by the other Maintainer (I think it was iGi or something like that) who has since been removed from AppCentral. And his Gitea version was stuck on 1.14.1.

Since upgrading to ADM4, I decided to migrate from his Gitea to yours while preserving all data. But even though the data migration worked, my repos failed to sync mainly because your version relied on the root user - which I have since manually fixed.

I am sure others who used iGi's Gitea will want to migrate to yours and they will also run into this problem.

**Comment by EndMove** [comment]
Hello nvatvani,
Can you give me more information about your repository problem?

Until version 1.15.5.r02 my application was using the "Gît" user and then I was notified by email of a setup problem with the ADM certificate and that's when I found the permission problem and changed the main Docker container user to root.
Nevertheless the change from "gît user" to "root user" went without problem during my tests and for this person who contacted me.

Since the application currently accesses data and certificates as root, such a permission problem should only occur when I change the main user to gît again (of course I would make sure that this does not happen by checking all permissions).

**Comment by nvatvani** [comment]
Aha ... interesting. It seems my first use of your gitea-adm was from 1.15.5.r02 where root user was used. I had no idea it was using the git user before.

It seems there are a few gitea-adm users in the Asustor world with conflicting setups requiring some to need Gitea being started as root, and others (who were previously using iGi's version) having the opposite requirement, i.e. Gitea being started as 'git' as-per guidelines set out by gitea.io.

I myself no longer have any issues with gitea-adm becasue my manual remediations (editing app.ini and changing docker env vars) to conform from the previous version maintained by iGi worked fine.

Let's wait and see what Asustor Support has to say.

**Comment by EndMove** [comment]
Hi nvatvani,

I understand completely and for the passage from git to Root here is the help I had written: https://github.com/EndMove/gitea-adm/wiki/HelpCenter#v1156r02-instructions-for-people-who-need-to-upgrade-to-version-

However, I have to inform you about another important problem that I just remembered! As you know Gitea normally uses within its Docker container a user named "git" with UID 1000 and a user group GID 1000. As you also know Gitea has direct access to your ASUSTOR server via "..Docker/Gitea/" (storage and configuration location).

The problem is the following: the user "git" and the group "git" do not exist in the Linux system of ADM, in fact the user UID 1000 in ADM is the first system user created during the setup of your server (in my case the user endmove) and the group GID will be the first group created

So in the case of a user not having created another user than "admin" and at least one additional group, his linux ADM system will be inconsistent because of the use of a non-existent UID and GID. This problem could cause a crash of ADM when calculating the daily space used per user.

In case the user does not have this first problem all gitea files and repositories will be allocated and put on the quota of the user UID 1000 and assigned to the group GID 1000. This compromises the security of Gitea and the integrity of the quota accounting for this user.

In conclusion, I would keep the user as root and make sure that the problem you had doesn't happen to anyone else in the next release. In addition, I strongly recommend that you restore the root user as the default Gitea user as well as recursively changing the permissions of the Gitea folder in Docker.

Sincerely,
Jérémi.

**Comment by EndMove** [comment]
Correction available in the last update, soon available from APP CENTRAL.
https://github.com/EndMove/gitea-adm/releases/tag/1.15.10.r01
